packages/nextjs/components/BetManager.tsx

```diff
@@ -1,8 +1,91 @@
-import { useEffect, useState } from "react";
+import { useEffect, useReducer, useState } from "react";
 import BetAmountPicker from "./BetAmountPicker";
 import HashPicker from "./HashPicker";
 import { cn } from "@/utils/cn";
 
+type BetAmountAction =
+  | { type: "SET_USD"; payload: number }
+  | { type: "SET_ETH"; payload: number }
+  | { type: "SET_WEI"; payload: number }
+  | { type: "SET_ETH_PRICE"; payload: number };
+
+export class BetAmount {
+  public usd: number;
+  public eth: number;
+  public wei: number;
+  public ethPrice: number;
+
+  constructor(ethPrice: number, usd = 0, eth = 0, wei = 0) {
+    this.ethPrice = ethPrice;
+    this.usd = usd;
+    this.eth = eth;
+    this.wei = wei;
+  }
+
+  setEthPrice(value: number) {
+    this.ethPrice = value;
+    this.usd = this.eth * value;
+  }
+
+  setUsd(value: number) {
+    this.usd = value;
+    this.eth = value / this.ethPrice;
+    this.wei = this.eth * 1e18;
+  }
+
+  setEth(value: number) {
+    this.eth = value;
+    this.usd = value * this.ethPrice;
+    this.wei = this.eth * 1e18;
+  }
+
+  setWei(value: number) {
+    this.wei = value;
+    this.eth = value / 1e18;
+    this.usd = this.eth * this.ethPrice;
+  }
+}
+
+type BetAmountState = BetAmount;
+
+const betAmountReducer = (state: BetAmountState, action: BetAmountAction): BetAmountState => {
+  switch (action.type) {
+    case "SET_USD":
+      const newUsdState = new BetAmount(state.ethPrice, action.payload, state.eth, state.wei);
+      newUsdState.setUsd(action.payload);
+      return newUsdState;
+    case "SET_ETH":
+      const newEthState = new BetAmount(state.ethPrice, state.usd, action.payload, state.wei);
+      newEthState.setEth(action.payload);
+      return newEthState;
+    case "SET_WEI":
+      const newWeiState = new BetAmount(state.ethPrice, state.usd, state.eth, action.payload);
+      newWeiState.setWei(action.payload);
+      return newWeiState;
+    case "SET_ETH_PRICE":
+      const newPriceState = new BetAmount(action.payload, state.usd, state.eth, state.wei);
+      newPriceState.setEthPrice(action.payload);
+      return newPriceState;
+    default:
+      throw new Error(`Unknown action type: ${action}`);
+  }
+};
+
+const useBetAmount = (initialEthPrice: number) => {
+  const [betAmount, dispatch] = useReducer<React.Reducer<BetAmount, BetAmountAction>>(
+    betAmountReducer,
+    new BetAmount(initialEthPrice),
+  );
+
+  return {
+    betAmount,
+    updateUsd: (usd: number) => dispatch({ type: "SET_USD", payload: usd }),
+    updateEth: (eth: number) => dispatch({ type: "SET_ETH", payload: eth }),
+    updateWei: (wei: number) => dispatch({ type: "SET_WEI", payload: wei }),
+    updateEthPrice: (ethPrice: number) => dispatch({ type: "SET_ETH_PRICE", payload: ethPrice }),
+  };
+};
+
```

Comment:
This is pretty cool. I haven't used this hook before I'm pretty sure its new in React 18 or 19. I like how you did this.